metastore/src/main/scala/com/gu/tableversions/metastore/Metastore.scala

```diff
@@ -35,15 +41,54 @@ object Metastore {
 
   final case class TableChanges(operations: List[TableOperation])
 
+  object TableChanges {
+    def apply(operations: TableOperation*): TableChanges = TableChanges(operations.toList)
+  }
+
   sealed trait TableOperation
 
   object TableOperation {
-    final case class AddPartition(partition: PartitionVersion) extends TableOperation
-    final case class UpdatePartitionVersion(partition: PartitionVersion) extends TableOperation
+    final case class AddPartition(partitionVersion: PartitionVersion) extends TableOperation
+    final case class UpdatePartitionVersion(partitionVersion: PartitionVersion) extends TableOperation
     final case class RemovePartition(partition: Partition) extends TableOperation
-    final case class UpdateTableLocation(tableLocation: URI, versionNumber: VersionNumber) extends TableOperation
+    final case class UpdateTableVersion(versionNumber: VersionNumber) extends TableOperation
   }
 
-  def computeChanges(current: TableVersion, target: TableVersion): TableChanges = ???
+  def computeChanges(oldVersion: TableVersion, newVersion: TableVersion): TableChanges = {
+
+    val operations = if (oldVersion.partitionVersions.map(_.partition) == List(Partition.snapshotPartition)) {
+      assert(newVersion.partitionVersions.map(_.partition) == List(Partition.snapshotPartition),
+             "Can't change table from snapshot table to partitioned")
+
+      if (oldVersion.partitionVersions.head.version != newVersion.partitionVersions.head.version)
```

Comment:
Is partitionVersions always garantueed to have a head?

Reply:
Yes, due to the assertion above.

That said, I don't really like how the code has worked out, treating snapshot tables as a special case which has a single empty version. I would like to try remodelling this explicitly, at least in most of the APIs, so we avoid these kinds of checks. For example making `TableVersion` an ADT with concrete cases for snapshot tables and partitioned tables.

Reply:
I agree, making explicit and impossible to compile would be best.